scripts/cve_check.py

```diff
@@ -0,0 +1,610 @@
+
+        product = pkgname
+        if pkgname in cve_products:
+            product = cve_products[pkgname]["product"]
+
+        cvesInRecord = "Yes"
+        if len(info) == 0:
+            cvesInRecord = "No"
+            issues = []
+        else:
+            issues = []
+            for cve in sorted(cves[pkgname]):
+                issues.append(cves[pkgname][cve])
+
+        data = {
+            "version": "1",
+            "package": [
+                {
+                    "name": pkginfo["source"],
+                    "binary package name": pkginfo["bin_pkgs"],
+                    "version": str(pkginfo["version"]),
+                    "products": [
+                        {
+                            "product": product,
+                            "cvesInRecord": cvesInRecord,
+                        },
+                    ],
+                },
+            ],
+            "issue": issues,
+        }
+
+        with open(filename, "w") as f:
+            json.dump(data, f)
+
+def main(args):
+    cve_data_dl_dir = f"{args.cve_dir}/downloads"
+    if not os.path.exists(cve_data_dl_dir):
+        os.makedirs(cve_data_dl_dir)
+
+    db_file = nvd_cve.update_nvd_db(cve_data_dl_dir, args.nvd_api_key)
+    if db_file is None:
+        print("Faied to fetch CVE database from NVD")
+        exit(1)
+
+    debian_cve_list = debian_cve.fetch_cve_data(cve_data_dl_dir)
+    if debian_cve_list is None:
+        print("Failed to fetch CVE data from Debian")
+        exit(1)
+
+    cip_kernel_sec_dir = kernel_cve.fetch_cip_kernel_sec(cve_data_dl_dir)
+    if cip_kernel_sec_dir is None:
+        print("Failed to fetch kernel-cip-sec")
+        exit(1)
+
+    installed_pkgs = debian_cve.parse_dpkg_status(args.dpkgstatus)
+    cve_products = get_cve_products(args.extra_cve_product)
+    cve_ignore_list = get_cve_ignore(args.extra_cve_check_ignore)
+
+    uniq_installed_pkgs = create_unique_package(installed_pkgs)
+    linux_kernel_src_dir = os.path.abspath(args.linux_kernel_src_dir)
+    cves = create_cves_info(db_file, debian_cve_list, uniq_installed_pkgs, installed_pkgs, args.debian_codename, cve_products, linux_kernel_src_dir, cip_kernel_sec_dir)
+
+    cves = update_ignored_cves_status(cves, cve_ignore_list)
+
+    output_dir = f"{args.cve_dir}/cve-check-result"
+
+    if not os.path.exists(output_dir):
+        os.makedirs(output_dir)
+
+    formats = args.output_format.split(",")
+    for fmt in formats:
+        if fmt == "text":
+            write_text(cves, output_dir, uniq_installed_pkgs)
+        elif fmt == "json":
+            write_json(cves, output_dir, uniq_installed_pkgs, cve_products)
+
+    print(f"CVE check finished. CVE check results are stored in {output_dir}")
+
+def parse_options():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("--cve-dir", dest="cve_dir", help="Directory for download and store CVE data, output cve check result",
+            metavar="CVEDIR", default="cve", required=True)
```

Comment:
Can we specify or dynamically detect default values for `--cve-dir`, `--dpkg-status`, `--linux-kernel-source-dir` when we are in a build directory?

Reply:
It's difficult to get kernel source directory and .dpkg_status file paths outside bitbake process.

Reply:
Is it not possible to use `bitbake -e` to get hints?

Reply:
unfortunately, bitbake -e need some image/recipe name (e.g. emlinux-image-base, linux-cip, and so on) to get information. So, script cannot decide deciwhat kind of name is correct or not.

Reply:
How about teaching the image name to the script?
It is annoying for normal users to be required to specify the location of the dpkg status or the kernel source.

Reply:
@hiraku-wfs I updated command line parameters. It get path to kernel source directory/.dpkg_status  file/downloads directory from bitbake <image name> -e option. Also remove cve-dir options from script to make usage simple. To run cve-check.py user need run source setup-emlinux <build dir> first.

New usage is following command line.

```
./repos/meta-emlinux/scripts/cve_check.py \
--image emlinux-image-base \
--output-format text,json
```

CVE check results are stored in $DEPLOY_DIR/cve/<IMAGE_FULL_NAME>.

```
build@2fdb25733595:~/work$ ls /home/build/work/build/tmp/deploy/cve/emlinux-image-base-emlinux-bookworm-generic-x86-64
acl                   coreutils_cve.json               findutils                     iproute2_cve.json   libffi_cve.json        libxcrypt_cve.json     p11-kit_cve.json         sysvinit
acl_cve.json          cpio                             findutils_cve.json            iptables            libgcrypt20            libzstd                pam                      sysvinit_cve.json
adduser               cpio_cve.json                    gcc-12                        iptables_cve.json   libgcrypt20_cve.json   libzstd_cve.json       pam_cve.json             tar
adduser_cve.json      cryptsetup                       gcc-12_cve.json               iputils             libgpg-error_cve.json  linux-base_cve.json    pcre2                    tar_cve.json
apparmor              cryptsetup_cve.json              gdbm_cve.json                 iputils_cve.json    libidn2                linux-cip-rt           pcre2_cve.json           tzdata_cve.json
apparmor_cve.json     cyrus-sasl2                      glibc                         json-c              libidn2_cve.json       linux-cip-rt_cve.json  perl                     usrmerge_cve.json
apt                   cyrus-sasl2_cve.json             glibc_cve.json                json-c_cve.json     libksba                lsb_cve.json           perl_cve.json            util-linux
apt_cve.json          dash                             gmp                           keyutils_cve.json   libksba_cve.json       lvm2                   pinentry                 util-linux_cve.json
argon2_cve.json       dash_cve.json                    gmp_cve.json                  klibc               libmd_cve.json         lvm2_cve.json          pinentry_cve.json        vim
attr_cve.json         db5.3                            gnupg2                        klibc_cve.json      libmnl_cve.json        lz4                    procps                   vim_cve.json
audit                 db5.3_cve.json                   gnupg2_cve.json               kmod_cve.json       libseccomp             lz4_cve.json           procps_cve.json          xxhash_cve.json
audit_cve.json        debconf_cve.json                 gnutls28                      krb5                libseccomp_cve.json    mawk_cve.json          readline                 xz-utils
base-files            debian-archive-keyring_cve.json  gnutls28_cve.json             krb5_cve.json       libselinux_cve.json    ncurses                readline_cve.json        xz-utils_cve.json
base-files_cve.json   debianutils_cve.json             grep                          libassuan_cve.json  libsemanage_cve.json   ncurses_cve.json       sed_cve.json             zlib
base-passwd           diffutils_cve.json               grep_cve.json                 libbpf              libsepol               nettle                 sensible-utils           zlib_cve.json
base-passwd_cve.json  dpkg                             gzip                          libbpf_cve.json     libsepol_cve.json      nettle_cve.json        sensible-utils_cve.json
bash                  dpkg_cve.json                    gzip_cve.json                 libbsd              libtasn1-6             npth_cve.json          shadow
bash_cve.json         e2fsprogs                        hostname_cve.json             libbsd_cve.json     libtasn1-6_cve.json    openldap               shadow_cve.json
bzip2                 e2fsprogs_cve.json               initramfs-tools               libcap2             libtirpc               openldap_cve.json      sqlite3
bzip2_cve.json        elfutils                         initramfs-tools_cve.json      libcap2_cve.json    libtirpc_cve.json      openssl                sqlite3_cve.json
cdebconf_cve.json     elfutils_cve.json                init-system-helpers_cve.json  libcap-ng_cve.json  libunistring_cve.json  openssl_cve.json       systemd
coreutils             emlinux-customization_cve.json   iproute2                      libffi              libxcrypt              p11-kit                systemd_cve.json
```

Reply:
I tried running the updated script, but I got the following error.

```
miracle@480b326b91e7:/work2/eml3-t5-202312/b1$ ../repos/meta-emlinux/scripts/cve_check.py --image-name emlinux-image-base
Initialize nvd cve database /work2/eml3-t5-202312/b1/downloads/CVE/nvd_cve_db.db
Update NVD CVE database
Traceback (most recent call last):
  File "/work2/eml3-t5-202312/b1/../repos/meta-emlinux/scripts/cve_check.py", line 612, in <module>
    main(parse_options())
  File "/work2/eml3-t5-202312/b1/../repos/meta-emlinux/scripts/cve_check.py", line 554, in main
    db_file = nvd_cve.update_nvd_db(cve_data_dl_dir, args.nvd_api_key)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/eml3-t5-202312/b1/../repos/meta-emlinux/scripts/lib/python/cve/nvd_cve.py", line 284, in update_nvd_db
    if not fetch_all_cves(conn, last_modified, nvd_api_key):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/eml3-t5-202312/b1/../repos/meta-emlinux/scripts/lib/python/cve/nvd_cve.py", line 182, in fetch_all_cves
    with open(os.path.join(CVE_DATABASE_DIR, 'cve_check'), 'a') as cve_f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'cve/cve_check'
```

Reply:
@hiraku-wfs thanks. Fixed it.